File: try codes/Bank.dart
import 'dart:async';
import 'dart:collection';
import 'dart:math';

void main() {
  final bankSimulation = BankSimulation();
  bankSimulation.start(
      'FCFS'); // Change this to 'FCFS' or 'SJF' to run different algorithms
}

class BankSimulation {
  final Queue<Customer> queue = Queue<Customer>();
  final List<Teller> tellers = List.generate(3, (index) => Teller(index + 1));
  bool simulationRunning = true;
  final Random random = Random();

  void start(String schedulingAlgorithm) {
    _startCustomerArrival();
    for (var teller in tellers) {
      switch (schedulingAlgorithm) {
        case 'FCFS':
          _startTellerServiceFCFS(teller);
          break;
        case 'SJF':
          _startTellerServiceSJF(teller);
          break;
        case 'RR':
          int quantum = 3; // Define the time quantum for Round Robin
          _startTellerServiceRR(teller, quantum);
          break;
        default:
          print('Invalid scheduling algorithm');
          return;
      }
    }
    _waitForTermination();
  }

  void _startCustomerArrival() {
    Timer.periodic(Duration(seconds: random.nextInt(3) + 1), (timer) {
      if (!simulationRunning) {
        timer.cancel();
        return;
      }
      final customer = Customer(random.nextInt(10) + 1);
      queue.add(customer);
      print('Customer ${customer.id} enters the Queue');
    });
  }

  void _startTellerServiceFCFS(Teller teller) {
    Timer.periodic(Duration(milliseconds: 500), (timer) async {
      if (!simulationRunning) {
        timer.cancel();
        return;
      }
      if (queue.isNotEmpty) {
        final customer = queue.removeFirst();
        print('Customer ${customer.id} is in Teller ${teller.id}');
        await Future.delayed(Duration(seconds: customer.serviceTime));
        print('Customer ${customer.id} leaves the Teller ${teller.id}');
      }
    });
  }

  void _startTellerServiceSJF(Teller teller) {
    Timer.periodic(Duration(milliseconds: 500), (timer) async {
      if (!simulationRunning) {
        timer.cancel();
        return;
      }
      if (queue.isNotEmpty) {
        final sortedQueue = queue.toList()
          ..sort((a, b) => a.serviceTime.compareTo(b.serviceTime));
        final customer = sortedQueue.first;
        queue.remove(customer);
        print(
            'Customer ${customer.id} (Shortest Job) is in Teller ${teller.id}');
        await Future.delayed(Duration(seconds: customer.serviceTime));
        print(
            'Customer ${customer.id} (Shortest Job) leaves the Teller ${teller.id}');
      }
    });
  }

  void _startTellerServiceRR(Teller teller, int quantum) {
    Timer.periodic(Duration(milliseconds: 500), (timer) async {
      if (!simulationRunning) {
        timer.cancel();
        return;
      }
      if (queue.isNotEmpty) {
        final customer = queue.removeFirst();
        final serveTime = min(quantum, customer.remainingTime);
        print(
            'Customer ${customer.id} (Round Robin) is in Teller ${teller.id}');
        await Future.delayed(Duration(seconds: serveTime));
        customer.remainingTime -= serveTime;
        if (customer.remainingTime > 0) {
          queue.add(customer);
          print(
              'Customer ${customer.id} (Round Robin) goes back to the Queue with ${customer.remainingTime} seconds remaining');
        } else {
          print(
              'Customer ${customer.id} (Round Robin) leaves the Teller ${teller.id}');
        }
      }
    });
  }

  void _waitForTermination() {
    Future.delayed(Duration(seconds: 60), () {
      simulationRunning = false;
      print('Simulation ended');
    });
  }
}

class Customer {
  static int _idCounter = 0;
  final int id;
  final int serviceTime;
  int remainingTime;

  Customer(this.serviceTime)
      : id = ++_idCounter,
        remainingTime = serviceTime;
}

class Teller {
  final int id;
  Teller(this.id);
}

File: try codes/bank.py
import random
import time
import matplotlib.pyplot as plt

class Customer:
    def __init__(self, id):
        self.id = id
        self.service_time = random.randint(1, 10)  # in seconds
        self.arrival_time = time.time()
        self.start_time = None
        self.end_time = None

class Teller:
    def __init__(self, id, queue, scheduling_algorithm):
        self.id = id
        self.queue = queue
        self.scheduling_algorithm = scheduling_algorithm

    def serve_customer(self):
        while True:
            if self.queue:
                if self.scheduling_algorithm == 'SJF':
                    customer = min(self.queue, key=lambda x: x.service_time)
                    self.queue.remove(customer)
                else:
                    customer = self.queue.pop(0)

                customer.start_time = time.time()
                print(f'Customer {customer.id} is in Teller {self.id}')
                time.sleep(customer.service_time)
                customer.end_time = time.time()
                print(f'Customer {customer.id} leaves the Teller {self.id}')
            else:
                time.sleep(0.1)

class BankSimulation:
    def __init__(self, scheduling_algorithm):
        self.queue_max_size = 10
        self.customers = []
        self.tellers = []
        self.scheduling_algorithm = scheduling_algorithm
        self.queue = []

        for i in range(3):
            self.tellers.append(Teller(i + 1, self.queue, self.scheduling_algorithm))

    def generate_customers(self):
        customer_id = 1
        while True:
            if len(self.queue) < self.queue_max_size:
                new_customer = Customer(customer_id)
                self.queue.append(new_customer)
                print(f'Customer {new_customer.id} enters the Queue')
                self.customers.append(new_customer)
                time.sleep(random.randint(1, 3))
                customer_id += 1
            else:
                print('Queue is FULL.')
                time.sleep(1)

    def start(self):
        import threading

        threading.Thread(target=self.generate_customers, daemon=True).start()

        for teller in self.tellers:
            threading.Thread(target=teller.serve_customer, daemon=True).start()

        time.sleep(30)

        self.calculate_and_plot_metrics()

    def calculate_and_plot_metrics(self):
        served_customers = [c for c in self.customers if c.end_time]
        avg_turnaround_time = sum(c.end_time - c.arrival_time for c in served_customers) / len(served_customers)
        avg_waiting_time = sum(c.start_time - c.arrival_time for c in served_customers) / len(served_customers)
        avg_response_time = sum(c.start_time - c.arrival_time for c in served_customers) / len(served_customers)

        labels = ['Average Turnaround Time', 'Average Waiting Time', 'Average Response Time']
        values = [avg_turnaround_time, avg_waiting_time, avg_response_time]

        plt.bar(labels, values, color=['red', 'blue', 'green'])
        plt.ylabel('Time (seconds)')
        plt.title('Bank Simulation Metrics')
        plt.show()

if __name__ == '__main__':
    print("Select Scheduling Algorithm:")
    print("1. First-Come-First-Serve (FCFS)")
    print("2. Shortest Job First (SJF)")
    print("3. Round Robin (RR)")

    choice = int(input() or '1')
    algorithm = {1: 'FCFS', 2: 'SJF', 3: 'RR'}.get(choice, 'FCFS')

    bank_simulation = BankSimulation(algorithm)
    bank_simulation.start()
